### ISSUE TITLE
Rule Change: `radix` rule should have better options

### ISSUE THREAD

**Issue by bubokko** [eslint/eslint#19916]
### What rule do you want to change?

radix

### What change do you want to make?

Implement suggestions

### How do you think the change should be implemented?

A new option

### Example code

```js
parseInt("0xff", undefined)
```

### What does the rule currently do for this code?

> Invalid radix parameter, must be an integer between 2 and 36.

### What will the rule do after it's changed?

Currently, `radix` rule has one option, which takes one of the two values:
- `always` - always require radix
- `as-needed` - providing `10` radix explicitly is forbidden

I have a feeling that a person who implemented this rule forgot that `10` is *not* a default radix. I love `radix` rule to make sure my radix is between 2 and 36, but neither `always` or `as-needed` satisfy my needs.

I think the option for `radix` rule could be something like this:

```typescript
{
    requireRadix: boolean
}
```

- `requireRadix: true` - requires providing radix, just like `always` option now, except it would also accept `undefined`. Currently, `always` option takes away the automatic radix feature of `parseInt()`.
- `requireRadix: false` - skipping radix is valid

Assuming that missing radix is `10` can lead to serious bugs, especially when we use user's input. If user types `0xf`, it's gonna evaluate as `15`, not `0`.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

**Comment by nzakas** [comment]
Thanks for the suggestion. To make sure we're all on the same page, ECMAScript 5 [specifies](https://262.ecma-international.org/5.1/#sec-15.1.2.2):

>  If radix is undefined or 0, it is assumed to be 10 except when the number begins with the character pairs 0x or 0X, in which case a radix of 16 is assumed.

So the default really is `0` and has been for a long time. Strings beginning with `0x` are an edge case where most folks will undoubtedly include the radix as 16. The more common error case this change in ES5 was meant to address was passing in values with leading zeros (i.e., `010`) that would be interpreted as octal.

So, I'm not sure it makes sense to change the current behavior.

Can you explain why you'd want to pass in `undefined` as the radix?

**Comment by bubokko** [comment]
I'd pass `undefined` if I really wanted to rely on the default behavior. It is explicit and it's a clear intention. I'd really love the rule to allow `2-36` and `undefined`, while NOT allowing 0. I agree that using automatic conversion is an edge use case, but still would be nice to have it, even if just for the sake of demonstrating `parseInt()`'s result.

Perhaps better idea for the option would be `allowAutoRadix: boolean`? It could be `false` by default and that would require using the `radix` parameter, but only as a 2-36 number. If `true`, it would allow skipping the `radix` or providing `undefined` for it.

If we assume that automatic conversion is obsolete, that makes the current `as-needed` implementation even more absurd, because not only it relies on that, but also falsely assume that the radix will be evaluated to `10` when we skip it.

What do you think?

**Comment by nzakas** [comment]
I'm not keen on expanding the scope of this rule much. The goal is to make sure people don't forget to use a radix. I don't think we should be doing too much beyond what it's already doing.

If we want to allow `undefined` as a valid option for this rule, then that opens the door for other values as well, because `parseInt()` ignores any invalid radix. That means `undefined` works the same as `null`, `0`, `{}`, etc. Essentially what we'd be saying is "ensure there is a radix but don't check the value." 

Again, I don't see a strong argument to change how `as-needed` works at this point. `10` is the default radix, and anyone who really cares will just use `always`.

@eslint/eslint-team I could use some other opinions on this.

**Comment by bubokko** [comment]
> Again, I don't see a strong argument to change how `as-needed` works at this point. `10` is the default radix, and anyone who really cares will just use `always`.

Except `10` is *not* the default. I think in this case we shouln't just assume that 0x strings were edge cases, because with user input, there's no such thing as an edge case in my opinion. `parseInt()` is a bot annoying when it comes to its interface, like many other legacy JS functions. I'd ask that we don't pretend `10` was the default though, as this assumption leads to unexpected behavior.

**Comment by mdjermanovic** [comment]
I think it's a very rare case that someone wants the default behavior of the `parseInt()` function. In such cases, an eslint-disable comment can be used. So, I think the `"always"` option (the default behavior of this rule) is fine as-is, and there's no need to change it or add a new option that would allow using the default behavior of the `parseInt()` function.

As for the `"as-needed"` option, I don't think it should exist, because radix argument `10` is not redundant. So, if we were to make any changes to this rule, I'd be in favor or removing this option.

**Comment by nzakas** [comment]
> So, if we were to make any changes to this rule, I'd be in favor or removing this option.

We can't really remove an existing option because that would break any existing configs that use it. We could, however, change `as-needed` to work the same as `always`, because if we're saying the radix is always needed then the option name still makes sense. I'd be okay with that.

Then the only other question on this issue is whether or not we want to make affordances for using `undefined` or some other value that indicates you want the default behavior. My feeling is that this is a case for a disable directive, but open to other opinions.

**Comment by nzakas** [comment]
@eslint/eslint-tsc thoughts?

**Comment by mdjermanovic** [comment]
> Then the only other question on this issue is whether or not we want to make affordances for using `undefined` or some other value that indicates you want the default behavior. My feeling is that this is a case for a disable directive, but open to other opinions.

I also think this is a case for a disable directive.

> > So, if we were to make any changes to this rule, I'd be in favor or removing this option.
> 
> We can't really remove an existing option because that would break any existing configs that use it. We could, however, change `as-needed` to work the same as `always`, because if we're saying the radix is always needed then the option name still makes sense. I'd be okay with that.

I think it would be confusing to change the option to do the opposite of what it was doing, and also confusing that the two options work the same. If we can't remove it, maybe we could mark it as deprecated?

**Comment by nzakas** [comment]
> I think it would be confusing to change the option to do the opposite of what it was doing, and also confusing that the two options work the same. If we can't remove it, maybe we could mark it as deprecated?

I don't think it's doing the opposite of what it was doing before. I'd say it's behaving in the way the name intends (the radix is needed). 

I'm not sure why we'd deprecate just that option if we're saying the option has a bug in it. 

Another approach would be to deprecate both options and explain that they both now behave like `"always"` so you don't need to specify an option anymore.

**Comment by mdjermanovic** [comment]
> Another approach would be to deprecate both options and explain that they both now behave like `"always"` so you don't need to specify an option anymore.

This sounds good to me.

I think this should be treated as a breaking change (for v10 perhaps?)

**Comment by nzakas** [comment]
Sorry, just was too swamped to circle back around to this. We did agree to lock down v10, but given that we should have discussed this and didn't, I think we can add it in.

**Comment by mdjermanovic** [comment]
I'm working on this.
